src/dune_engine/fs_memo.ml

```diff
@@ -2,39 +2,93 @@ open! Stdune
 open! Import
 open Memo.Build.O
 
+(* Files and directories have non-overlapping sets of paths, so we can track
+   them using the same memoization table. *)
 let memo =
-  Memo.create "fs_notify" ~doc:"fs notify"
+  Memo.create "fs_memo" ~doc:"tracking files and directories on the file system"
     ~input:(module Path)
     ~output:(Simple (module Unit))
     ~visibility:Hidden
     (fun _path -> Memo.Build.return ())
 
-(* Currently, we do not expose this low-level primitive. If you need it, perhaps
-   you could add a higher-level primitive instead, such as [file_exists]? *)
+(* Declare a dependency on a path. Instead of calling [depend] directly, you
+   should prefer using the helper function [declaring_dependency], because it
+   calls [depend] and uses the corresponding path in the right order. *)
 let depend path =
   if Path.is_in_build_dir path then
     Code_error.raise "Fs_memo.depend called on a build path" [];
   Memo.exec memo path
 
-let file_exists path =
-  let+ () = depend path in
-  Path.exists path
+(* Declare a dependency on [path] and then sample the current value using the
+   supplied function [f].
 
-let file_digest path =
+   If the order is reversed, the value can change after we sample it but before
+   we register the dependency, which can result in memoizing a stale value. This
+   scenario is purely hypothetical (at least for now) but it's nice to rule it
+   out explicitly by doing things in the right order.
+
+   Currently, we do not expose this low-level primitive. If you need it, perhaps
+   you could add a higher-level primitive instead, such as [file_exists]? *)
+let declaring_dependency path ~f =
   let+ () = depend path in
-  Cached_digest.source_or_external_file path
+  f path
+
+let file_exists = declaring_dependency ~f:Path.exists
+
+(* CR-someday amokhov: It is unclear if we got the layers of abstraction right
+   here. One could argue that caching is a higher-level concept compared to file
+   watching, and we should expose this function from the [Cached_digest] module
+   instead. For now, we keep it here because it seems nice to group all tracked
+   file system access functions in one place, and exposing an uncached version
+   of [file_digest] seems error-prone. We may need to rethink this decision. *)
+let file_digest = declaring_dependency ~f:Cached_digest.source_or_external_file
+
+let dir_contents = declaring_dependency ~f:Path.readdir_unsorted_with_kinds
```

Comment:
`dir_contents` seems unlikely to work correctly, both with current and with "theoretically-ideal" events, because it's waiting for a directory event, but we're planning to have events for files inside it, instead.

As an aside, I don't really see how to review the event kinds set you've written down since we neither have anything that uses them nor anything that can produce them.

Reply:
Yes, I think it makes sense to add the logic that handles different kinds of event into this PR. To make `dir_contents dir` work, my plan is to invalidate it in the following two cases:

* When receiving `Directory_deleted` event whose `path` is equal to `dir`.
* When receiving `File_created` and `File_deleted` events whose parent path is equal to `dir`.

Does this make sense?

Reply:
Similarly, a file path should be invalidated in the following cases:
* When receiving `File_deleted` event with the matching path.
* When receiving `File_changed` event with the matching path. 
* Receiving `File_created` event for a tracked path should be impossible, but it probably makes sense to still invalidate the path in this case, as opposed to crashing or ignoring the event.

Reply:
And I should probably also list my assumptions about the backend:
* When a directory is deleted, the backend generates `File_deleted` and `Directory_deleted` events for the whole tree.
* When a directory is renamed, the backend generates `File_deleted`, `File_created`, `Directory_created` and `Directory_deleted` events for the whole tree.

Reply:
I just looked at [inotifywait manual](https://linux.die.net/man/1/inotifywait), and here are the kinds of events it can produce:

* `modify`: a watched file or a file within a watched directory was written to. This gives us `File_changed`.
* `attrib`: the metadata of a watched file or a file within a watched directory was modified. This includes timestamps, file permissions, extended attributes etc. We should map it to `File_changed` as well.
* `create`: a file or directory was created within a watched directory. This gives us `X_created` and we don't immediately know if `X=File` or `X=Directory`. But in fact this distinction doesn't seem to be important for us, so perhaps I can collapse these two event into one.
* `delete`: a file or directory within a watched directory was deleted. As above, this gives us `X_deleted`.
* `moved_to`: a file or directory was moved into a watched directory. This event occurs even if the file is simply moved from and to the same directory. This gives us `X_created` events, potentially for a whole directory, recursively.
* `moved_from`: a file or directory was moved from a watched directory. This event occurs even if the file is simply moved from and to the same directory. This gives us `X_deleted` events, potentially for a whole directory, recursively.

There are a few other events, but I think we should focus on the above for now.

Reply:
Yeah, OK, sounds like we'll be able to do it with inotifywait.

Reply:
OK, I pushed a commit that implements a basic event-handling strategy and documents some of the above discussion.